prompt the use of `--nocapture` flag if `cargo test` process is terminated via a signal.

stupendoussuperpowers opened the pull request:
<!-- homu-ignore:start -->
<!--
NOTICE: Due to limited review capacity, the Cargo team is not accepting new
features or major changes at this time. Please consult with the team before
opening a new PR. Only issues that have been explicitly marked as accepted
will be reviewed.

Thanks for submitting a pull request 🎉! Here are some tips for you:

* If this is your first contribution, read "Cargo Contribution Guide":
  https://doc.crates.io/contrib/
* Run `cargo fmt --all` to format your code changes.
* Small commits and pull requests are always preferable and easy to review.
* If your idea is large and needs feedback from the community, read how:
  https://doc.crates.io/contrib/process/#working-on-large-features
* Cargo takes care of compatibility. Read our design principles:
  https://doc.crates.io/contrib/design.html
* When changing help text of cargo commands, follow the steps to generate docs:
  https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages
* If your PR is not finished, set it as "draft" PR or add "WIP" in its title.
* It's ok to use the CI resources to test your PR, but please don't abuse them.

### What does this PR try to resolve?

Explain the motivation behind this change.
A clear overview along with an in-depth explanation are helpful.

You can use `Fixes #<issue number>` to associate this PR to an existing issue.

### How should we test and review this PR?

Demonstrate how you test this change and guide reviewers through your PR.
With a smooth review process, a pull request usually gets reviewed quicker.

If you don't know how to write and run your tests, please read the guide:
https://doc.crates.io/contrib/tests

### Additional information

Other information you want to mention in this PR, such as prior arts,
future extensions, an unresolved problem, or a TODO list.
-->
<!-- homu-ignore:end -->

Fixes #10855 

As per the discussion on this issue, we want to prompt the user to use `--nocapture` if a test is terminated abnormally. The motivation for this change is described in the issue.

We check for 3 things before we display this flag. - 
- `!is_simple` (if the test ended with a non 101 status code)
- `harness` (if the standard test harness was used), and
- `!nocapture` (whether or not the `--nocapture` flag was already passed to the test)

There's further tests added to `test::nonzero_exit_status` that check that the `stderr` is correct for the various combinations possible when a test ends with a non-101 status code.

The new expected behavior is - 
- Display `--nocapture` note for only non-zero exit statuses, when the `--nocapture` flag is not passed. 
- Only display the note if we use a standard test harness since custom test harnesses do not implement the `--nocapture` flag.

To implement the check for the `--nocapture` flag, the function definition for `report_test_errors` was changed to add the `test_args: &[&str]` parameter. This parameter is passed from the immediate calling function. This private function is only called twice change and is not causing regression after making the appropriate changes to both the places it's called in.

epage commented on src/cargo/ops/cargo_test.rs:
```diff
@@ -431,9 +431,6 @@ fn report_test_error(
     crate::display_error(&err, &mut ws.config().shell());
 
-    // println!("{:#?}", opts);
-    // println!("{:#?}", test_args);
```
Could you squash this commit?

epage commented on tests/testsuite/test.rs:
```diff
@@ -4792,6 +4792,21 @@ error: test failed, to rerun pass `--test t1`
+        .with_status(4)
+        .run();
+
+    p.cargo("test --test t2 -- --nocapture")
```
Can you split these new test cases out into a commit (first in the PR).  It should pass, showing the behavior of `master`.  The commit with all of the other changes should then update the test to still pass.  This will make it easier to see (1) the tests work and (2) what behavior changed

epage commented on src/cargo/ops/cargo_test.rs:
```diff
@@ -429,4 +430,13 @@ fn report_test_error(
+
+    if !is_simple && harness && !nocapture {
+        drop(ws.config().shell().note(
+            "test was terminated via a signal. stderr might be truncated. use `--nocapture` flag to disable output buffering."
```
I feel like this is long enough that its going to lose users (and technically a test can exist from something besides a signal, like `std::process::exit`)

How about

> "test exited abnormally; to see the full output pass `--nocapture` to the harness"

stupendoussuperpowers commented:
I've split the changes into 2 commits as you asked, and changed the display string. 

The tests are failing right now because it is outputting the --nocapture note even if there's an exec failure. Looking into a fix for this.

ehuss commented:
r? epage

Hope you don't mind taking this.

epage commented on tests/testsuite/test.rs:
```diff
@@ -4792,6 +4792,35 @@ error: test failed, to rerun pass `--test t1`
+        .with_status(4)
+        .run();
+
+    p.cargo("test --test t2 -- --nocapture")
```
Looks like there was a merge problem and the second commit re-adds the test cases the first commit added.

stupendoussuperpowers replied:
Let me take a look at all these commits once, they've gotten a bit confusing. Will push with correct changes in a bit.

stupendoussuperpowers commented:
Hi @epage need some help with this. 

`cargo test -p cargo` for `linux x86_64 beta` is passing, but the test fails for `linux x86_64 nightly.` Looks like the `--no-fail-fast --nocapture` are outputting different things for these two. Not sure how to deal with this.

epage commented:
> Hi @epage need some help with this.
> 
> `cargo test -p cargo` for `linux x86_64 beta` is passing, but the test fails for `linux x86_64 nightly.` Looks like the `--no-fail-fast --nocapture` are outputting different things for these two. Not sure how to deal with this.

`[..]` are a type of wildcard.  You can adjust these wildcards to make it still show the relevant information (`t1.rs`) while hiding the information that is tied to a specific rust version

epage commented:
@stupendoussuperpowers could you rebase, instead of merge, so I can more easily browse the individual commits?

stupendoussuperpowers commented:
give me a day or so to work on this, really embarrassing but i'm running into multiple git issues on my local machine. will mark it ready once it's done.

epage commented:
btw I'm hosting office hours right now but also available at other times in case you want to do a call.  See https://github.com/rust-lang/cargo/wiki/Office-Hours

epage commented:
Hmm, forgot to look back at this.  @stupendoussuperpowers anything left or is this good to go?

epage commented:
@bors r+

Thanks for all the work you did on this!

bors commented:
:pushpin: Commit dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4 has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4 -->
<!-- homu: {"type":"Approved","sha":"dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4 with merge 4312a90e8e5105505b33455dc621367c86f93284...
<!-- homu: {"type":"BuildStarted","head_sha":"dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4","merge_sha":"4312a90e8e5105505b33455dc621367c86f93284"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/5825528604/job/15797948393)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/actions/runs/5825528604/job/15797948393","builder_name":"checks-actions"} -->

weihanglo commented:
```
 thread 'git_auth::ssh_something_happens' panicked at '
test failed running `/home/runner/work/cargo/cargo/target/debug/cargo check -v`
error: expected to find:
[..]Connection [..] by [..]

did not find in output:
    Updating git repository `ssh://127.0.0.1:37377/foo/bar`
warning: spurious network error (3 tries remaining): failed to fill whole buffer
error: failed to get `bar` as a dependency of package `foo v0.0.1 (/home/runner/work/cargo/cargo/target/tmp/cit/t99/foo)`

Caused by:
  failed to load source for dependency `bar`

Caused by:
  Unable to update ssh://127.0.0.1:37377/foo/bar

Caused by:
  failed to clone into: /home/runner/work/cargo/cargo/target/tmp/cit/t99/home/.cargo/git/db/bar-d4c725e0d0110fc9

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  An IO error occurred when talking to the server

Caused by:
  ssh: connect to host 127.0.0.1 port 37377: Connection refused
', tests/testsuite/git_auth.rs:301:10
```

Not sure what's going on. Could be the TCP connection accidentally closed.

Let's @bors retry and see.

bors commented:
:hourglass: Testing commit dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4 with merge 7da103076ee73578411dda0d7997d24dddf39ffa...
<!-- homu: {"type":"BuildStarted","head_sha":"dd8cd9c9a93db2a362cc4ec719a4226e21c43bd4","merge_sha":"7da103076ee73578411dda0d7997d24dddf39ffa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/5826221551/job/15800904012)
Approved by: epage
Pushing 7da103076ee73578411dda0d7997d24dddf39ffa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/5826221551/job/15800904012"},"merge_sha":"7da103076ee73578411dda0d7997d24dddf39ffa"} -->
